C++ API parity: MultiheadAttention

pbelevich opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#27309 C++ API parity: MultiheadAttention**

Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

pbelevich committed:
C++ API parity: MultiheadAttention

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"




[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

yf225 commented on torch/csrc/api/include/torch/nn/functional/activation.h:
```diff
@@ -65,6 +66,95 @@ inline Tensor rrelu(Tensor& input, const RReLUOptions& options, bool training=fa
+  const c10::optional<Tensor>& k_proj_weight = c10::nullopt,
+  const c10::optional<Tensor>& v_proj_weight = c10::nullopt,
+  const c10::optional<Tensor>& static_k = c10::nullopt,
+  const c10::optional<Tensor>& static_v = c10::nullopt) {
```
I feel that we should have another options called `MultiheadAttentionForwardOptions` to represent all the optional arguments for this functional, and also use that options in the module's forward method to represent the optional arguments. Do you think it would be a good idea?

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

yf225 commented on torch/csrc/api/include/torch/nn/functional/activation.h:
```diff
@@ -81,6 +82,95 @@ inline Tensor softplus(const Tensor& input, const SoftplusOptions& options) {
+  int64_t num_heads,
+  const Tensor& in_proj_weight,
+  const Tensor& in_proj_bias,
+  const c10::optional<Tensor>& bias_k,
```
I think we should also use `const Tensor&` to represent an optional tensor (its null value would be the undefined tensor `Tensor()`), which is consistent with how optional tensor is represented elsewhere in the C++ API (e.g. custom autograd functions).

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

yf225 reviewed (commented):
@pbelevich Thanks so much for the awesome work! I left some initial comments.

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

anjali411 commented on torch/csrc/api/src/nn/options/activation.cpp:
```diff
@@ -29,7 +29,18 @@ SoftminFuncOptions::SoftminFuncOptions(int64_t dim) : dim_(dim) {}
+    Tensor bias_k, Tensor bias_v,
+    bool add_zero_attn, double dropout_p,
+    Tensor out_proj_weight, Tensor out_proj_bias
+  ) : embed_dim_to_check_(embed_dim_to_check), num_heads_(num_heads),
```
I think you can fix the clang-tidy failures by doing:
``embed_dim_to_check_(std::move(embed_dim_to_check)), num_heads_(std::move(num_heads))``

anjali411 commented on torch/csrc/api/src/nn/options/activation.cpp:
```diff
@@ -29,7 +29,18 @@ SoftminFuncOptions::SoftminFuncOptions(int64_t dim) : dim_(dim) {}
+    bool add_zero_attn, double dropout_p,
+    Tensor out_proj_weight, Tensor out_proj_bias
+  ) : embed_dim_to_check_(embed_dim_to_check), num_heads_(num_heads),
+      in_proj_weight_(in_proj_weight), in_proj_bias_(in_proj_bias),
```
``in_proj_weight_(std::move(in_proj_weight)), in_proj_bias_(std::move(in_proj_bias))`` here

anjali411 commented on torch/csrc/api/src/nn/options/activation.cpp:
```diff
@@ -29,7 +29,18 @@ SoftminFuncOptions::SoftminFuncOptions(int64_t dim) : dim_(dim) {}
+    Tensor out_proj_weight, Tensor out_proj_bias
+  ) : embed_dim_to_check_(embed_dim_to_check), num_heads_(num_heads),
+      in_proj_weight_(in_proj_weight), in_proj_bias_(in_proj_bias),
+      bias_k_(bias_k), bias_v_(bias_v),
```
``bias_k_(std::move(bias_k)), bias_v_(std::move(bias_v))``

anjali411 commented on torch/csrc/api/src/nn/options/activation.cpp:
```diff
@@ -29,7 +29,18 @@ SoftminFuncOptions::SoftminFuncOptions(int64_t dim) : dim_(dim) {}
+  ) : embed_dim_to_check_(embed_dim_to_check), num_heads_(num_heads),
+      in_proj_weight_(in_proj_weight), in_proj_bias_(in_proj_bias),
+      bias_k_(bias_k), bias_v_(bias_v),
+      add_zero_attn_(add_zero_attn), dropout_p_(dropout_p),
```
``add_zero_attn_(std::move(add_zero_attn)), dropout_p_(std::move(dropout_p))``

anjali411 commented on torch/csrc/api/src/nn/options/activation.cpp:
```diff
@@ -29,7 +29,18 @@ SoftminFuncOptions::SoftminFuncOptions(int64_t dim) : dim_(dim) {}
+      in_proj_weight_(in_proj_weight), in_proj_bias_(in_proj_bias),
+      bias_k_(bias_k), bias_v_(bias_v),
+      add_zero_attn_(add_zero_attn), dropout_p_(dropout_p),
+      out_proj_weight_(out_proj_weight), out_proj_bias_(out_proj_bias) {}
```
``out_proj_weight_(std::move(out_proj_weight)), out_proj_bias_(std::move(out_proj_bias))

anjali411 commented on torch/csrc/api/include/torch/nn/modules/activation.h:
```diff
@@ -507,5 +508,45 @@ class TORCH_API ThresholdImpl : public torch::nn::Cloneable<ThresholdImpl> {
+  Tensor q_proj_weight;
+  Tensor k_proj_weight;
+  Tensor v_proj_weight;
+  int head_dim;
```
I think head_dim should be int64_t

yf225 commented on torch/csrc/api/include/torch/nn/options/activation.h:
```diff
@@ -315,5 +315,102 @@ struct TORCH_API GumbelSoftmaxFuncOptions {
+    }
+    if (!vdim_) {
+      vdim_ = embed_dim;
+    }
```
I believe `kdim_` and `vdim_` are always equal to `c10::nullopt` when the constructor is run, and we can probably make the type of `kdim_` and `vdim_` non-optional, and always assign them the same value as `embed_dim` in the  constructor instead.

yf225 commented on torch/csrc/api/src/nn/modules/activation.cpp:
```diff
@@ -390,5 +391,116 @@ void ThresholdImpl::pretty_print(std::ostream& stream) const {
+  const Tensor& query, const Tensor& key,
+  const Tensor& value, const Tensor& key_padding_mask,
+  bool need_weights, const Tensor& attn_mask) {
+  if (_qkv_same_embed_dim) {
```
In the Python version, the first if-branch seems to be `if not self._qkv_same_embed_dim:`, is there a reason we switch the order?

yf225 commented on torch/csrc/api/src/nn/modules/activation.cpp:
```diff
@@ -390,5 +391,116 @@ void ThresholdImpl::pretty_print(std::ostream& stream) const {
+  if (!_qkv_same_embed_dim) {
+    q_proj_weight = torch::empty({options.embed_dim(), options.embed_dim()});
+    k_proj_weight = torch::empty({options.embed_dim(), *options.kdim()});
+    v_proj_weight = torch::empty({options.embed_dim(), *options.vdim()});
```
we probably need to add the equivalent of `self.register_parameter('in_proj_weight', None)` here, to mirror the Python version better

yf225 commented on torch/csrc/api/src/nn/modules/activation.cpp:
```diff
@@ -390,5 +391,116 @@ void ThresholdImpl::pretty_print(std::ostream& stream) const {
+    k_proj_weight = torch::empty({options.embed_dim(), *options.kdim()});
+    v_proj_weight = torch::empty({options.embed_dim(), *options.vdim()});
+  } else {
+    in_proj_weight = torch::empty({3 * options.embed_dim(), options.embed_dim()});
```
we probably need to add the equivalent of
```python
self.register_parameter('q_proj_weight', None)
self.register_parameter('k_proj_weight', None)
self.register_parameter('v_proj_weight', None)
```
here, to mirror the Python version better

yf225 commented on torch/csrc/api/src/nn/modules/activation.cpp:
```diff
@@ -390,5 +391,116 @@ void ThresholdImpl::pretty_print(std::ostream& stream) const {
+  } else {
+    bias_k = {};
+    bias_v = {};
+  }
```
I believe all of the logic above should be in the `reset()` function instead, and all of the logic below should stay in the `_reset_parameters()` function :D

yf225 commented on torch/csrc/api/src/nn/modules/activation.cpp:
```diff
@@ -390,5 +391,116 @@ void ThresholdImpl::pretty_print(std::ostream& stream) const {
+}
+
+void MultiheadAttentionImpl::pretty_print(std::ostream& stream) const {
+  stream << "torch::nn::MultiheadAttention(todo)";
```
`todo` :3

yf225 commented on test/cpp/api/modules.cpp:
```diff
@@ -3044,6 +3050,379 @@ TEST_F(ModulesTest, BCEWithLogitsLoss) {
+    /*same_embed_dim=*/true
+  );
+}
+
```
Thanks so much for the thorough tests!

yf225 commented on torch/csrc/api/include/torch/nn/options/activation.h:
```diff
@@ -315,5 +315,102 @@ struct TORCH_API GumbelSoftmaxFuncOptions {
+namespace functional {
+
+/// Options for `torch::nn::functional::multi_head_attention_forward`
+struct TORCH_API MultiheadAttentionForwardOptions {
```
We might want to change it to `MultiheadAttentionForwardFuncOptions`, to be more consistent with the naming scheme of other functional options :D

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit b5e5cc4:
* **18/18** failures introduced in this PR



## Detailed failure analysis

One may explore the probable reasons each build failed interactively [on the Dr. CI website](https://dr.pytorch.org/commit-details.html?sha1=b5e5cc47ddf180b5aa2d05ecfaa297d667a3ad30).

### 17 new failures recognized by patterns
The following build failures don't appear to be due to upstream breakage:
#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956366) caffe2_onnx_py3_6_clang7_ubuntu16_04_test (1/17)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=74494720))
<details>
<summary>
<code>Dec 15 03:03:33 ../.local/lib/python3.6/site-packages/caffe2/python/allcompare_test.py::TestAllCompare::test_allcompare FAILED [  0%]</code>
</summary>

```
Dec 15 03:03:27 ============================= test session starts ============================== 
Dec 15 03:03:27 platform linux -- Python 3.6.9, pytest-5.2.2, py-1.8.0, pluggy-0.13.0 -- /usr/bin/python3.6 
Dec 15 03:03:27 cachedir: .pytest_cache 
Dec 15 03:03:27 rootdir: /var/lib/jenkins 
Dec 15 03:03:27 plugins: hypothesis-3.59.0, sugar-0.9.2 
Dec 15 03:03:27 collecting ... WARNING:root:This caffe2 python run does not have GPU support. Will run in CPU only mode. 
Dec 15 03:03:27 INFO:caffe2.python.net_drawer:Cannot import pydot, which is required for drawing a network. This can usually be installed in python with "pip install pydot". Also, pydot requires graphviz to convert dot files to pdf: in ubuntu, this can usually be installed with "sudo apt-get install graphviz". 
Dec 15 03:03:33 net_drawer will not run correctly. Please install the correct dependencies. 
Dec 15 03:03:33 collected 2934 items / 1 skipped / 2933 selected 
Dec 15 03:03:33  
Dec 15 03:03:33 ../.local/lib/python3.6/site-packages/caffe2/python/allcompare_test.py::TestAllCompare::test_allcompare FAILED [  0%] 
Dec 15 03:03:33  
Dec 15 03:03:33 =================================== FAILURES =================================== 
Dec 15 03:03:33 ________________________ TestAllCompare.test_allcompare ________________________ 
Dec 15 03:03:33  
Dec 15 03:03:33 self = <caffe2.python.allcompare_test.TestAllCompare testMethod=test_allcompare> 
Dec 15 03:03:33  
Dec 15 03:03:33     @given( 
Dec 15 03:03:33 >       d=st.integers(1, 5), n=st.integers(2, 11), num_procs=st.integers(1, 8) 
Dec 15 03:03:33     ) 
Dec 15 03:03:33     def test_allcompare(self, d, n, num_procs): 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956372) caffe2_onnx_py2_gcc5_ubuntu16_04_test (2/17)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=74495705))
<details>
<summary>
<code>Dec 15 03:06:44 ../.local/lib/python2.7/site-packages/caffe2/python/allcompare_test.py::TestAllCompare::test_allcompare FAILED [  0%]</code>
</summary>

```
Dec 15 03:06:36 ============================= test session starts ============================== 
Dec 15 03:06:36 platform linux2 -- Python 2.7.12, pytest-4.6.6, py-1.8.0, pluggy-0.13.0 -- /usr/bin/python2 
Dec 15 03:06:36 cachedir: .pytest_cache 
Dec 15 03:06:36 rootdir: /var/lib/jenkins 
Dec 15 03:06:36 plugins: hypothesis-3.59.0, sugar-0.9.2 
Dec 15 03:06:36 collecting ... WARNING:root:This caffe2 python run does not have GPU support. Will run in CPU only mode. 
Dec 15 03:06:36 INFO:caffe2.python.net_drawer:Cannot import pydot, which is required for drawing a network. This can usually be installed in python with "pip install pydot". Also, pydot requires graphviz to convert dot files to pdf: in ubuntu, this can usually be installed with "sudo apt-get install graphviz". 
Dec 15 03:06:44 net_drawer will not run correctly. Please install the correct dependencies. 
Dec 15 03:06:44 collected 2934 items / 1 skipped / 2933 selected 
Dec 15 03:06:44  
Dec 15 03:06:44 ../.local/lib/python2.7/site-packages/caffe2/python/allcompare_test.py::TestAllCompare::test_allcompare FAILED [  0%] 
Dec 15 03:06:44  
Dec 15 03:06:44 =================================== FAILURES =================================== 
Dec 15 03:06:44 ________________________ TestAllCompare.test_allcompare ________________________ 
Dec 15 03:06:44  
Dec 15 03:06:44 self = <caffe2.python.allcompare_test.TestAllCompare testMethod=test_allcompare> 
Dec 15 03:06:44  
Dec 15 03:06:44     @given( 
Dec 15 03:06:44 >       d=st.integers(1, 5), n=st.integers(2, 11), num_procs=st.integers(1, 8) 
Dec 15 03:06:44     ) 
Dec 15 03:06:44     def test_allcompare(self, d, n, num_procs): 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956382) pytorch_linux_xenial_py2_7_9_test (3/17)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=74495761))
<details>
<summary>
<code>Dec 15 03:06:29 caused by: Connection refused (os error 111)</code>
</summary>

```
Dec 15 03:06:29 +++ eval 'extract_trap_cmd ' 
Dec 15 03:06:29 ++++ extract_trap_cmd 
Dec 15 03:06:29 ++++ printf '%s\n' '' 
Dec 15 03:06:29 +++ printf '%s\n' cleanup 
Dec 15 03:06:29 ++ trap -- ' 
Dec 15 03:06:29 cleanup' EXIT 
Dec 15 03:06:29 ++ which sccache 
Dec 15 03:06:29 ++ sccache --stop-server 
Dec 15 03:06:29 Stopping sccache server... 
Dec 15 03:06:29 error: couldn't connect to server 
Dec 15 03:06:29 caused by: Connection refused (os error 111) 
Dec 15 03:06:29 ++ true 
Dec 15 03:06:29 ++ rm /var/lib/jenkins/sccache_error.log 
Dec 15 03:06:29 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Dec 15 03:06:29 ++ SCCACHE_IDLE_TIMEOUT=1200 
Dec 15 03:06:29 ++ RUST_LOG=sccache::server=error 
Dec 15 03:06:29 ++ sccache --start-server 
Dec 15 03:06:29 Starting sccache server... 
Dec 15 03:06:29 ++ sccache --zero-stats 
Dec 15 03:06:29 Compile requests                 0 
Dec 15 03:06:29 Compile requests executed        0 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956373) pytorch_linux_xenial_py3_6_gcc5_4_test (4/17)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=74495814))
<details>
<summary>
<code>Dec 15 03:07:05 caused by: Connection refused (os error 111)</code>
</summary>

```
Dec 15 03:07:05 +++ eval 'extract_trap_cmd ' 
Dec 15 03:07:05 ++++ extract_trap_cmd 
Dec 15 03:07:05 ++++ printf '%s\n' '' 
Dec 15 03:07:05 +++ printf '%s\n' cleanup 
Dec 15 03:07:05 ++ trap -- ' 
Dec 15 03:07:05 cleanup' EXIT 
Dec 15 03:07:05 ++ which sccache 
Dec 15 03:07:05 ++ sccache --stop-server 
Dec 15 03:07:05 Stopping sccache server... 
Dec 15 03:07:05 error: couldn't connect to server 
Dec 15 03:07:05 caused by: Connection refused (os error 111) 
Dec 15 03:07:05 ++ true 
Dec 15 03:07:05 ++ rm /var/lib/jenkins/sccache_error.log 
Dec 15 03:07:05 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Dec 15 03:07:05 ++ SCCACHE_IDLE_TIMEOUT=1200 
Dec 15 03:07:05 ++ RUST_LOG=sccache::server=error 
Dec 15 03:07:05 ++ sccache --start-server 
Dec 15 03:07:05 Starting sccache server... 
Dec 15 03:07:05 ++ sccache --zero-stats 
Dec 15 03:07:05 Compile requests                 0 
Dec 15 03:07:05 Compile requests executed        0 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956383) pytorch_macos_10_13_py3_test (5/17)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=74495952))
<details>
<summary>
<code>Dec 14 19:08:56 RuntimeError: test_cpp_extensions failed!</code>
</summary>

```
Dec 14 19:08:56 /Users/distiller/workspace/miniconda3/lib/python3.7/site-packages/torch/include/torch/csrc/api/include/torch/nn/functional/activation.h:10:10: fatal error: 'ATen/native/cpu/Loops.h' file not found 
Dec 14 19:08:56 #include <ATen/native/cpu/Loops.h> 
Dec 14 19:08:56          ^~~~~~~~~~~~~~~~~~~~~~~~~ 
Dec 14 19:08:56 1 error generated. 
Dec 14 19:08:56 error: command 'clang' failed with exit status 1 
Dec 14 19:08:56 Traceback (most recent call last): 
Dec 14 19:08:56   File "test/run_test.py", line 456, in <module> 
Dec 14 19:08:56     main() 
Dec 14 19:08:56   File "test/run_test.py", line 449, in main 
Dec 14 19:08:56     raise RuntimeError(message) 
Dec 14 19:08:56 RuntimeError: test_cpp_extensions failed! 
Dec 14 19:08:56 + cleanup 
Dec 14 19:08:56 + retcode=1 
Dec 14 19:08:56 + set +x 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956287) pytorch_xla_linux_xenial_py3_6_clang7_build (6/17)
**Step:** "Build" ([details](https://dr.pytorch.org/build-details.html?build_id=74496534))
<details>
<summary>
<code>Dec 15 02:43:59 caused by: Connection refused (os error 111)</code>
</summary>

```
Dec 15 02:43:59 +++ eval 'extract_trap_cmd ' 
Dec 15 02:43:59 ++++ extract_trap_cmd 
Dec 15 02:43:59 ++++ printf '%s\n' '' 
Dec 15 02:43:59 +++ printf '%s\n' cleanup 
Dec 15 02:43:59 ++ trap -- ' 
Dec 15 02:43:59 cleanup' EXIT 
Dec 15 02:43:59 ++ which sccache 
Dec 15 02:43:59 ++ sccache --stop-server 
Dec 15 02:43:59 Stopping sccache server... 
Dec 15 02:43:59 error: couldn't connect to server 
Dec 15 02:43:59 caused by: Connection refused (os error 111) 
Dec 15 02:43:59 ++ true 
Dec 15 02:43:59 ++ rm /var/lib/jenkins/sccache_error.log 
Dec 15 02:43:59 rm: cannot remove '/var/lib/jenkins/sccache_error.log': No such file or directory 
Dec 15 02:43:59 ++ true 
Dec 15 02:43:59 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Dec 15 02:43:59 ++ SCCACHE_IDLE_TIMEOUT=1200 
Dec 15 02:43:59 ++ RUST_LOG=sccache::server=error 
Dec 15 02:43:59 ++ sccache --start-server 
Dec 15 02:43:59 Starting sccache server... 
Dec 15 02:43:59 ++ sccache --zero-stats 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956396) pytorch_linux_xenial_py3_clang5_asan_test (7/17)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=74499555))
<details>
<summary>
<code>Dec 15 03:16:46 caused by: Connection refused (os error 111)</code>
</summary>

```
Dec 15 03:16:46 +++ eval 'extract_trap_cmd ' 
Dec 15 03:16:46 ++++ extract_trap_cmd 
Dec 15 03:16:46 ++++ printf '%s\n' '' 
Dec 15 03:16:46 +++ printf '%s\n' cleanup 
Dec 15 03:16:46 ++ trap -- ' 
Dec 15 03:16:46 cleanup' EXIT 
Dec 15 03:16:46 ++ which sccache 
Dec 15 03:16:46 ++ sccache --stop-server 
Dec 15 03:16:46 Stopping sccache server... 
Dec 15 03:16:46 error: couldn't connect to server 
Dec 15 03:16:46 caused by: Connection refused (os error 111) 
Dec 15 03:16:46 ++ true 
Dec 15 03:16:46 ++ rm /var/lib/jenkins/sccache_error.log 
Dec 15 03:16:46 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Dec 15 03:16:46 ++ SCCACHE_IDLE_TIMEOUT=1200 
Dec 15 03:16:46 ++ RUST_LOG=sccache::server=error 
Dec 15 03:16:46 ++ sccache --start-server 
Dec 15 03:16:46 Starting sccache server... 
Dec 15 03:16:46 ++ sccache --zero-stats 
Dec 15 03:16:46 Compile requests                 0 
Dec 15 03:16:46 Compile requests executed        0 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956429) pytorch_windows_test1 (8/17)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=74500171))
<details>
<summary>
<code> - cmake (exited 1) - cmake not installed. An error occurred during installation:</code>
</summary>

```
 Unable to resolve dependency 'cmake.install (= 3.16.1)'. 
cmake package files install completed. Performing other installation steps. 
The install of cmake was NOT successful. 
cmake not installed. An error occurred during installation: 
 Unable to resolve dependency 'cmake.install (= 3.16.1)'. 
 
Chocolatey installed 0/1 packages. 1 packages failed. 
 See the log for details (C:\ProgramData\chocolatey\logs\chocolatey.log). 
 
Failures 
 - cmake (exited 1) - cmake not installed. An error occurred during installation: 
 Unable to resolve dependency 'cmake.install (= 3.16.1)'. 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956424) pytorch_linux_xenial_cuda9_cudnn7_py3_NO_AVX_NO_AVX2_test (9/17)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=74500625))
<details>
<summary>
<code>Dec 15 03:21:51 caused by: Connection refused (os error 111)</code>
</summary>

```
Dec 15 03:21:51 +++ eval 'extract_trap_cmd ' 
Dec 15 03:21:51 ++++ extract_trap_cmd 
Dec 15 03:21:51 ++++ printf '%s\n' '' 
Dec 15 03:21:51 +++ printf '%s\n' cleanup 
Dec 15 03:21:51 ++ trap -- ' 
Dec 15 03:21:51 cleanup' EXIT 
Dec 15 03:21:51 ++ which sccache 
Dec 15 03:21:51 ++ sccache --stop-server 
Dec 15 03:21:51 Stopping sccache server... 
Dec 15 03:21:51 error: couldn't connect to server 
Dec 15 03:21:51 caused by: Connection refused (os error 111) 
Dec 15 03:21:51 ++ true 
Dec 15 03:21:51 ++ rm /var/lib/jenkins/sccache_error.log 
Dec 15 03:21:51 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Dec 15 03:21:51 ++ SCCACHE_IDLE_TIMEOUT=1200 
Dec 15 03:21:51 ++ RUST_LOG=sccache::server=error 
Dec 15 03:21:51 ++ sccache --start-server 
Dec 15 03:21:51 Starting sccache server... 
Dec 15 03:21:51 ++ sccache --zero-stats 
Dec 15 03:21:51 Compile requests                 0 
Dec 15 03:21:51 Compile requests executed        0 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956420) pytorch_linux_xenial_cuda9_cudnn7_py3_test (10/17)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=74500741))
<details>
<summary>
<code>Dec 15 03:21:59 caused by: Connection refused (os error 111)</code>
</summary>

```
Dec 15 03:21:59 +++ eval 'extract_trap_cmd ' 
Dec 15 03:21:59 ++++ extract_trap_cmd 
Dec 15 03:21:59 ++++ printf '%s\n' '' 
Dec 15 03:21:59 +++ printf '%s\n' cleanup 
Dec 15 03:21:59 ++ trap -- ' 
Dec 15 03:21:59 cleanup' EXIT 
Dec 15 03:21:59 ++ which sccache 
Dec 15 03:21:59 ++ sccache --stop-server 
Dec 15 03:21:59 Stopping sccache server... 
Dec 15 03:21:59 error: couldn't connect to server 
Dec 15 03:21:59 caused by: Connection refused (os error 111) 
Dec 15 03:21:59 ++ true 
Dec 15 03:21:59 ++ rm /var/lib/jenkins/sccache_error.log 
Dec 15 03:21:59 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Dec 15 03:21:59 ++ SCCACHE_IDLE_TIMEOUT=1200 
Dec 15 03:21:59 ++ RUST_LOG=sccache::server=error 
Dec 15 03:21:59 ++ sccache --start-server 
Dec 15 03:21:59 Starting sccache server... 
Dec 15 03:21:59 ++ sccache --zero-stats 
Dec 15 03:21:59 Compile requests                 0 
Dec 15 03:21:59 Compile requests executed        0 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956417) pytorch_linux_xenial_cuda9_cudnn7_py3_nogpu_test (11/17)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=74501151))
<details>
<summary>
<code>Dec 15 03:24:02 caused by: Connection refused (os error 111)</code>
</summary>

```
Dec 15 03:24:02 +++ eval 'extract_trap_cmd ' 
Dec 15 03:24:02 ++++ extract_trap_cmd 
Dec 15 03:24:02 ++++ printf '%s\n' '' 
Dec 15 03:24:02 +++ printf '%s\n' cleanup 
Dec 15 03:24:02 ++ trap -- ' 
Dec 15 03:24:02 cleanup' EXIT 
Dec 15 03:24:02 ++ which sccache 
Dec 15 03:24:02 ++ sccache --stop-server 
Dec 15 03:24:02 Stopping sccache server... 
Dec 15 03:24:02 error: couldn't connect to server 
Dec 15 03:24:02 caused by: Connection refused (os error 111) 
Dec 15 03:24:02 ++ true 
Dec 15 03:24:02 ++ rm /var/lib/jenkins/sccache_error.log 
Dec 15 03:24:02 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Dec 15 03:24:02 ++ SCCACHE_IDLE_TIMEOUT=1200 
Dec 15 03:24:02 ++ RUST_LOG=sccache::server=error 
Dec 15 03:24:02 ++ sccache --start-server 
Dec 15 03:24:02 Starting sccache server... 
Dec 15 03:24:02 ++ sccache --zero-stats 
Dec 15 03:24:02 Compile requests                 0 
Dec 15 03:24:02 Compile requests executed        0 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956423) pytorch_linux_xenial_cuda9_cudnn7_py3_slow_test (12/17)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=74501250))
<details>
<summary>
<code>Dec 15 03:22:45 caused by: Connection refused (os error 111)</code>
</summary>

```
Dec 15 03:22:45 +++ eval 'extract_trap_cmd ' 
Dec 15 03:22:45 ++++ extract_trap_cmd 
Dec 15 03:22:45 ++++ printf '%s\n' '' 
Dec 15 03:22:45 +++ printf '%s\n' cleanup 
Dec 15 03:22:45 ++ trap -- ' 
Dec 15 03:22:45 cleanup' EXIT 
Dec 15 03:22:45 ++ which sccache 
Dec 15 03:22:45 ++ sccache --stop-server 
Dec 15 03:22:45 Stopping sccache server... 
Dec 15 03:22:45 error: couldn't connect to server 
Dec 15 03:22:45 caused by: Connection refused (os error 111) 
Dec 15 03:22:45 ++ true 
Dec 15 03:22:45 ++ rm /var/lib/jenkins/sccache_error.log 
Dec 15 03:22:45 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Dec 15 03:22:45 ++ SCCACHE_IDLE_TIMEOUT=1200 
Dec 15 03:22:45 ++ RUST_LOG=sccache::server=error 
Dec 15 03:22:45 ++ sccache --start-server 
Dec 15 03:22:45 Starting sccache server... 
Dec 15 03:22:45 ++ sccache --zero-stats 
Dec 15 03:22:45 Compile requests                 0 
Dec 15 03:22:45 Compile requests executed        0 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956422) pytorch_linux_xenial_cuda9_cudnn7_py3_NO_AVX2_test (13/17)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=74501463))
<details>
<summary>
<code>Dec 15 03:23:25 caused by: Connection refused (os error 111)</code>
</summary>

```
Dec 15 03:23:25 +++ eval 'extract_trap_cmd ' 
Dec 15 03:23:25 ++++ extract_trap_cmd 
Dec 15 03:23:25 ++++ printf '%s\n' '' 
Dec 15 03:23:25 +++ printf '%s\n' cleanup 
Dec 15 03:23:25 ++ trap -- ' 
Dec 15 03:23:25 cleanup' EXIT 
Dec 15 03:23:25 ++ which sccache 
Dec 15 03:23:25 ++ sccache --stop-server 
Dec 15 03:23:25 Stopping sccache server... 
Dec 15 03:23:25 error: couldn't connect to server 
Dec 15 03:23:25 caused by: Connection refused (os error 111) 
Dec 15 03:23:25 ++ true 
Dec 15 03:23:25 ++ rm /var/lib/jenkins/sccache_error.log 
Dec 15 03:23:25 ++ SCCACHE_ERROR_LOG=/var/lib/jenkins/sccache_error.log 
Dec 15 03:23:25 ++ SCCACHE_IDLE_TIMEOUT=1200 
Dec 15 03:23:25 ++ RUST_LOG=sccache::server=error 
Dec 15 03:23:25 ++ sccache --start-server 
Dec 15 03:23:25 Starting sccache server... 
Dec 15 03:23:25 ++ sccache --zero-stats 
Dec 15 03:23:25 Compile requests                 0 
Dec 15 03:23:25 Compile requests executed        0 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956414) binary_linux_libtorch_2_7m_cpu_gcc5_4_cxx11-abi_shared-with-deps_test (14/17)
**Step:** "Run in docker" ([details](https://dr.pytorch.org/build-details.html?build_id=74501693))
<details>
<summary>
<code>Dec 15 03:27:53 /tmp/libtorch/include/torch/csrc/api/include/torch/nn/functional/activation.h:10:35: fatal error: ATen/native/cpu/Loops.h: No such file or directory</code>
</summary>

```
Dec 15 03:27:53 + TORCH_CPU_LINK_FLAGS= 
Dec 15 03:27:53 + '[' -f /tmp/libtorch/lib/libtorch_cpu.so ']' 
Dec 15 03:27:53 + TORCH_CPU_LINK_FLAGS=-ltorch_cpu 
Dec 15 03:27:53 + g++ example-app.cpp -I/tmp/libtorch/include -I/tmp/libtorch/include/torch/csrc/api/include -D_GLIBCXX_USE_CXX11_ABI=1 -std=gnu++14 -L/tmp/libtorch/lib -Wl,-R/tmp/libtorch/lib -ltorch -ltorch_cpu -lc10 -o example-app 
Dec 15 03:27:53 In file included from /tmp/libtorch/include/torch/csrc/api/include/torch/nn/functional/loss.h:4:0, 
Dec 15 03:27:53                  from /tmp/libtorch/include/torch/csrc/api/include/torch/nn/functional.h:10, 
Dec 15 03:27:53                  from /tmp/libtorch/include/torch/csrc/api/include/torch/nn.h:4, 
Dec 15 03:27:53                  from /tmp/libtorch/include/torch/csrc/api/include/torch/all.h:7, 
Dec 15 03:27:53                  from /tmp/libtorch/include/torch/csrc/api/include/torch/torch.h:3, 
Dec 15 03:27:53                  from example-app.cpp:1: 
Dec 15 03:27:53 /tmp/libtorch/include/torch/csrc/api/include/torch/nn/functional/activation.h:10:35: fatal error: ATen/native/cpu/Loops.h: No such file or directory 
Dec 15 03:27:53 compilation terminated. 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956426) binary_linux_libtorch_2_7m_cpu_devtoolset7_shared-with-deps_test (15/17)
**Step:** "Run in docker" ([details](https://dr.pytorch.org/build-details.html?build_id=74501741))
<details>
<summary>
<code>Dec 15 03:28:15 /tmp/libtorch/include/torch/csrc/api/include/torch/nn/functional/activation.h:10:10: fatal error: ATen/native/cpu/Loops.h: No such file or directory</code>
</summary>

```
Dec 15 03:28:14 + TORCH_CPU_LINK_FLAGS= 
Dec 15 03:28:14 + '[' -f /tmp/libtorch/lib/libtorch_cpu.so ']' 
Dec 15 03:28:14 + TORCH_CPU_LINK_FLAGS=-ltorch_cpu 
Dec 15 03:28:14 + g++ example-app.cpp -I/tmp/libtorch/include -I/tmp/libtorch/include/torch/csrc/api/include -D_GLIBCXX_USE_CXX11_ABI=0 -std=gnu++14 -L/tmp/libtorch/lib -Wl,-R/tmp/libtorch/lib -ltorch -ltorch_cpu -lc10 -o example-app 
Dec 15 03:28:15 In file included from /tmp/libtorch/include/torch/csrc/api/include/torch/nn/functional/loss.h:4:0, 
Dec 15 03:28:15                  from /tmp/libtorch/include/torch/csrc/api/include/torch/nn/functional.h:10, 
Dec 15 03:28:15                  from /tmp/libtorch/include/torch/csrc/api/include/torch/nn.h:4, 
Dec 15 03:28:15                  from /tmp/libtorch/include/torch/csrc/api/include/torch/all.h:7, 
Dec 15 03:28:15                  from /tmp/libtorch/include/torch/csrc/api/include/torch/torch.h:3, 
Dec 15 03:28:15                  from example-app.cpp:1: 
Dec 15 03:28:15 /tmp/libtorch/include/torch/csrc/api/include/torch/nn/functional/activation.h:10:10: fatal error: ATen/native/cpu/Loops.h: No such file or directory 
Dec 15 03:28:15  #include <ATen/native/cpu/Loops.h> 
Dec 15 03:28:15           ^~~~~~~~~~~~~~~~~~~~~~~~~ 
Dec 15 03:28:15 compilation terminated. 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956421) pytorch_python_doc_push (16/17)
**Step:** "Doc Build and Push" ([details](https://dr.pytorch.org/build-details.html?build_id=74501790))
<details>
<summary>
<code>Dec 15 03:28:27 /opt/conda/lib/python3.6/site-packages/torch/include/torch/csrc/api/include/torch/nn/functional/activation.h:10:35: fatal error: ATen/native/cpu/Loops.h: No such file or directory</code>
</summary>

```
ackages/torch/include -I/opt/conda/lib/python3.6/site-packages/torch/include/torch/csrc/api/include -I/opt/conda/lib/python3.6/site-packages/torch/include/TH -I/opt/conda/lib/python3.6/site-packages/torch/include/THC -I/opt/conda/include/python3.6m -c /var/lib/jenkins/workspace/vision/torchvision/csrc/vision.cpp -o build/temp.linux-x86_64-3.6/var/lib/jenkins/workspace/vision/torchvision/csrc/vision.o -DTORCH_API_INCLUDE_EXTENSION_H -DTORCH_EXTENSION_NAME=_C -D_GLIBCXX_USE_CXX11_ABI=1 -std=c++14 
Dec 15 03:28:27 cc1plus: warning: command line option '-Wstrict-prototypes' is valid for C/ObjC but not for C++ 
Dec 15 03:28:27 In file included from /opt/conda/lib/python3.6/site-packages/torch/include/torch/csrc/api/include/torch/nn/functional/loss.h:4:0, 
Dec 15 03:28:27                  from /opt/conda/lib/python3.6/site-packages/torch/include/torch/csrc/api/include/torch/nn/functional.h:10, 
Dec 15 03:28:27                  from /opt/conda/lib/python3.6/site-packages/torch/include/torch/csrc/api/include/torch/nn.h:4, 
Dec 15 03:28:27                  from /opt/conda/lib/python3.6/site-packages/torch/include/torch/csrc/api/include/torch/all.h:7, 
Dec 15 03:28:27                  from /opt/conda/lib/python3.6/site-packages/torch/include/torch/extension.h:4, 
Dec 15 03:28:27                  from /var/lib/jenkins/workspace/vision/torchvision/csrc/cpu/vision_cpu.h:2, 
Dec 15 03:28:27                  from /var/lib/jenkins/workspace/vision/torchvision/csrc/DeformConv.h:3, 
Dec 15 03:28:27                  from /var/lib/jenkins/workspace/vision/torchvision/csrc/vision.cpp:8: 
Dec 15 03:28:27 /opt/conda/lib/python3.6/site-packages/torch/include/torch/csrc/api/include/torch/nn/functional/activation.h:10:35: fatal error: ATen/native/cpu/Loops.h: No such file or directory 
Dec 15 03:28:27 compilation terminated. 
Dec 15 03:28:27 error: command 'gcc' failed with exit status 1 
```

</details>

#### [![See CircleCI build](https://avatars0.githubusercontent.com/ml/7?s=12)](https://circleci.com/gh/pytorch/pytorch/3956272) binary_macos_libtorch_2_7_cpu_build (17/17)
**Step:** "Test" ([details](https://dr.pytorch.org/build-details.html?build_id=74508997))
<details>
<summary>
<code>/tmp/libtorch/include/torch/csrc/api/include/torch/nn/functional/activation.h:10:10: fatal error: 'ATen/native/cpu/Loops.h' file not found</code>
</summary>

```
+ '[' -f /tmp/libtorch/lib/libtorch_cpu.so ']' 
+ '[' -f /tmp/libtorch/lib/libtorch_cpu.dylib ']' 
+ TORCH_CPU_LINK_FLAGS=-ltorch_cpu 
+ g++ example-app.cpp -I/tmp/libtorch/include -I/tmp/libtorch/include/torch/csrc/api/include -D_GLIBCXX_USE_CXX11_ABI=0 -std=gnu++14 -L/tmp/libtorch/lib -Wl,-rpath /tmp/libtorch/lib -ltorch -ltorch_cpu -lc10 -o example-app 
In file included from example-app.cpp:1: 
In file included from /tmp/libtorch/include/torch/csrc/api/include/torch/torch.h:3: 
In file included from /tmp/libtorch/include/torch/csrc/api/include/torch/all.h:7: 
In file included from /tmp/libtorch/include/torch/csrc/api/include/torch/nn.h:4: 
In file included from /tmp/libtorch/include/torch/csrc/api/include/torch/nn/functional.h:10: 
In file included from /tmp/libtorch/include/torch/csrc/api/include/torch/nn/functional/loss.h:4: 
/tmp/libtorch/include/torch/csrc/api/include/torch/nn/functional/activation.h:10:10: fatal error: 'ATen/native/cpu/Loops.h' file not found 
#include <ATen/native/cpu/Loops.h> 
         ^~~~~~~~~~~~~~~~~~~~~~~~~ 
1 error generated. 
```

</details>

### 1 failure *not* recognized by patterns:
| Job | Step | Status |
| --- | --- | --- |
| ![CircleCI](https://avatars0.githubusercontent.com/ml/7?s=12) <sup>pytorch_windows_test2</sup> | <sup>Test</sup> | New in PR |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

anjali411 commented on test/cpp/api/modules.cpp:
```diff
@@ -3044,6 +3050,379 @@ TEST_F(ModulesTest, BCEWithLogitsLoss) {
+  namespace F = torch::nn::functional;
+
+  torch::Tensor _batchmatmul(const torch::Tensor& a, const torch::Tensor& b) {
+    assert(a.size(0) == b.size(0));
```
should we use TORCH_INTERNAL_ASSERT instead?
``TORCH_INTERNAL_ASSERT(a.size(0) == b.size(0));``

anjali411 commented on test/cpp/api/modules.cpp:
```diff
@@ -3044,6 +3050,379 @@ TEST_F(ModulesTest, BCEWithLogitsLoss) {
+
+  torch::Tensor _batchmatmul(const torch::Tensor& a, const torch::Tensor& b) {
+    assert(a.size(0) == b.size(0));
+    assert(a.size(1) == b.size(1));
```
same question as above

anjali411 commented on torch/csrc/api/include/torch/nn/options/activation.h:
```diff
@@ -315,5 +315,96 @@ struct TORCH_API GumbelSoftmaxFuncOptions {
+
+/// Options for MultiheadAttention functional and module.
+struct TORCH_API MultiheadAttentionOptions {
+  MultiheadAttentionOptions(int64_t embed_dim, int64_t num_heads)
```
I think we should move this constructor to options/activation.cpp for consistency

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

anjali411 commented on torch/csrc/api/include/torch/nn/options/activation.h:
```diff
@@ -315,5 +315,96 @@ struct TORCH_API GumbelSoftmaxFuncOptions {
+  TORCH_ARG(bool, add_zero_attn) = false;
+
+  /// total number of features in key. Default: c10::nullopt.
+  TORCH_ARG(int64_t, kdim);
```
where is it declared as an optional variable? I am thinking if we should change it to:
``TORCH_ARG(c10::optional<int64_t>, kdim) = c10::nullopt;``

pbelevich replied:
https://github.com/pytorch/pytorch/pull/27309#discussion_r356274944

anjali411 commented on torch/csrc/api/include/torch/nn/options/activation.h:
```diff
@@ -315,5 +315,96 @@ struct TORCH_API GumbelSoftmaxFuncOptions {
+  TORCH_ARG(int64_t, kdim);
+
+  /// total number of features in key. Default: c10::nullopt.
+  TORCH_ARG(int64_t, vdim);
```
same as above
``TORCH_ARG(c10::optional<int64_t>, vdim) = c10::nullopt;``

pbelevich replied:
https://github.com/pytorch/pytorch/pull/27309#discussion_r356274944

pbelevich committed:
Update on "C++ API parity: MultiheadAttention"


Differential Revision: [D17766736](https://our.internmc.facebook.com/intern/diff/D17766736)

[ghstack-poisoned]

anjali411 commented on torch/csrc/api/src/nn/modules/activation.cpp:
```diff
@@ -390,5 +391,119 @@ void ThresholdImpl::pretty_print(std::ostream& stream) const {
+}
+
+void MultiheadAttentionImpl::reset() {
+  _qkv_same_embed_dim = options.kdim() == options.embed_dim() &&
```
and after making kdim and vdim optional in options/activation.h, we can add code corresponding to these lines in activation.py
self.kdim = kdim if kdim is not None else embed_dim
self.vdim = vdim if vdim is not None else embed_dim
``if(options.kdim() != c10::nullopt) {
      options.kdim(options.embed_dim());
    }``
``if(options.vdim() != c10::nullopt) {       
options.vdim(options.embed_dim());
}``

anjali411 reviewed (approved):
LGTM :)
